File: packages/cordova-plugin/ios/OSFileTransferPlugin.swift
import Foundation
import Combine
import IONFileTransferLib
import QuartzCore

private enum Action: String {
    case download
    case upload
}

@objc(OSFileTransferPlugin)
class OSFileTransferPlugin : CDVPlugin {
    private lazy var manager = IONFLTRManager()
    private var listeners: [CDVInvokedUrlCommand] = []
    private var cancellables = Set<AnyCancellable>()
    private var lastProgressReportTime = CACurrentMediaTime()
    private let progressUpdateInterval: TimeInterval = 0.1 // 100ms

    @objc(downloadFile:)
    func downloadFile(command: CDVInvokedUrlCommand) {
        do {
            let (serverURL, fileURL, shouldTrackProgress, httpOptions) = try validateAndPrepare(command: command, action: .download)

            try manager.downloadFile(
                fromServerURL: serverURL,
                toFileURL: fileURL,
                withHttpOptions: httpOptions
            ).sink(
                receiveCompletion: handleCompletion(command: command, source: serverURL.absoluteString, target: fileURL.absoluteString),
                receiveValue: handleReceiveValue(
                    command: command,
                    type: .download,
                    url: serverURL.absoluteString,
                    path: fileURL.path,
                    shouldTrackProgress: shouldTrackProgress
                )
            ).store(in: &cancellables)
        } catch {
            sendError(command, error.toFileTransferError(),
                      source: getParameter(command, "url") as? String,
                      target: getParameter(command, "path") as? String)
        }
    }

    @objc(uploadFile:)
    func uploadFile(command: CDVInvokedUrlCommand) {
        do {
            let (serverURL, fileURL, shouldTrackProgress, httpOptions) = try validateAndPrepare(command: command, action: .upload)

            let chunkedMode = getBoolParameter(command, "chunkedMode") ?? false
            let mimeType = getParameter(command, "mimeType") as? String
            let fileKey = getParameter(command, "fileKey") as? String ?? "file"

            let uploadOptions = IONFLTRUploadOptions(
                chunkedMode: chunkedMode,
                mimeType: mimeType,
                fileKey: fileKey
            )

            try manager.uploadFile(
                fromFileURL: fileURL,
                toServerURL: serverURL,
                withUploadOptions: uploadOptions,
                andHttpOptions: httpOptions
            ).sink(
                receiveCompletion: handleCompletion(command: command, source: fileURL.absoluteString, target: serverURL.absoluteString),
                receiveValue: handleReceiveValue(
                    command: command,
                    type: .upload,
                    url: serverURL.absoluteString,
                    path: fileURL.path,
                    shouldTrackProgress: shouldTrackProgress
                )
            ).store(in: &cancellables)
        } catch {
            sendError(command, error.toFileTransferError(),
                      source: getParameter(command, "path") as? String,
                      target: getParameter(command, "url") as? String)
        }
    }

    @objc(addListener:)
    func addListener(command: CDVInvokedUrlCommand) {
        listeners.append(command)
    }

    @objc(removeAllListeners:)
    func removeAllListeners(command: CDVInvokedUrlCommand) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK)
        commandDelegate.send(result, callbackId: command.callbackId)

        for listener in listeners {
            let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "removeAllListeners was called")
            commandDelegate.send(result, callbackId: listener.callbackId)
        }

        listeners.removeAll()
    }

    private func validateAndPrepare(command: CDVInvokedUrlCommand, action: Action) throws -> (URL, URL, Bool, IONFLTRHttpOptions) {
        guard let options = command.arguments[0] as? [String: Any] else {
            throw OSFileTransferError.invalidParameters()
        }

        guard let urlString = options["url"] as? String, !urlString.isEmpty else {
            throw OSFileTransferError.invalidServerUrl(nil)
        }

        guard let serverURL = URL(string: urlString) else {
            throw OSFileTransferError.invalidServerUrl(urlString)
        }

        guard let filePath = options["path"] as? String, !filePath.isEmpty else {
            throw OSFileTransferError.invalidParameters("Path is required.")
        }

        // ✅ FIXED: Use proper file URL constructor
        let fileURL = URL(fileURLWithPath: filePath)

        let shouldTrackProgress = options["progress"] as? Bool ?? false

        let httpOptions = createHttpOptions(from: options, defaultMethod: defaultHTTPMethod(for: action))

        return (serverURL, fileURL, shouldTrackProgress, httpOptions)
    }

    private func defaultHTTPMethod(for action: Action) -> String {
        switch action {
        case .download: return "GET"
        case .upload: return "POST"
        }
    }

    private func createHttpOptions(from options: [String: Any], defaultMethod: String) -> IONFLTRHttpOptions {
        let method = options["method"] as? String ?? defaultMethod
        let headers = options["headers"] as? [String: String] ?? [:]
        let params = extractParams(from: options["params"] as? [String: Any] ?? [:])
        let timeoutInSeconds = (options["connectTimeout"] as? Int ?? 60000) / 1000
        let disableRedirects = options["disableRedirects"] as? Bool ?? false
        let shouldEncodeUrlParams = options["shouldEncodeUrlParams"] as? Bool ?? true

        return IONFLTRHttpOptions(
            method: method,
            params: params,
            headers: headers,
            timeout: Int(timeoutInSeconds),
            disableRedirects: disableRedirects,
            shouldEncodeUrlParams: shouldEncodeUrlParams
        )
    }

    private func extractParams(from params: [String: Any]) -> [String: [String]] {
        var result: [String: [String]] = [:]

        for (key, value) in params {
            if let stringValue = value as? String {
                result[key] = [stringValue]
            } else if let arrayValue = value as? [Any] {
                let stringArray = arrayValue.compactMap { $0 as? String }
                if !stringArray.isEmpty {
                    result[key] = stringArray
                }
            }
        }

        return result
    }

    private func handleCompletion(command: CDVInvokedUrlCommand, source: String, target: String) -> (Subscribers.Completion<Error>) -> Void {
        return { completion in
            if case let .failure(error) = completion {
                self.sendError(command, error.toFileTransferError(), source: source, target: target)
            }
        }
    }

    private func handleReceiveValue(
        command: CDVInvokedUrlCommand,
        type: Action,
        url: String,
        path: String,
        shouldTrackProgress: Bool
    ) -> (IONFLTRTransferResult) -> Void {
        return { result in
            switch result {
            case .ongoing(let status):
                self.reportProgressIfNeeded(
                    type: type,
                    url: url,
                    bytes: status.bytes,
                    contentLength: status.contentLength,
                    lengthComputable: status.lengthComputable,
                    shouldTrack: shouldTrackProgress
                )

            case .complete(let data):
                self.reportProgressIfNeeded(
                    type: type,
                    url: url,
                    bytes: data.totalBytes,
                    contentLength: data.totalBytes,
                    lengthComputable: true,
                    shouldTrack: shouldTrackProgress,
                    force: true
                )

                let resultDict: [String: Any] = {
                    switch type {
                    case .download:
                        return ["path": path]
                    case .upload:
                        return [
                            "bytesSent": data.totalBytes,
                            "responseCode": data.responseCode,
                            "response": data.responseBody ?? "",
                            "headers": data.headers
                        ]
                    }
                }()

                self.sendSuccess(command, resultDict)
            }
        }
    }

    private func reportProgressIfNeeded(
        type: Action,
        url: String,
        bytes: Int,
        contentLength: Int,
        lengthComputable: Bool,
        shouldTrack: Bool,
        force: Bool = false
    ) {
        guard shouldTrack else { return }

        let current = CACurrentMediaTime()
        guard force || (current - lastProgressReportTime >= progressUpdateInterval) else { return }
        lastProgressReportTime = current

        let progressData: [String: Any] = [
            "type": type.rawValue,
            "url": url,
            "bytes": bytes,
            "contentLength": contentLength,
            "lengthComputable": lengthComputable
        ]

        for listener in listeners {
            let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: progressData)
            result?.keepCallback = true
            commandDelegate.send(result, callbackId: listener.callbackId)
        }
    }

    private func sendSuccess(_ command: CDVInvokedUrlCommand, _ result: [String: Any] = [:]) {
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: result)
        commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    private func sendError(_ command: CDVInvokedUrlCommand, _ error: OSFileTransferError, source: String? = nil, target: String? = nil) {
        var errorWithContext = error
        errorWithContext.source = source
        errorWithContext.target = target

        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: errorWithContext.toDictionary())
        commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    private func getParameter(_ command: CDVInvokedUrlCommand, _ key: String) -> Any? {
        guard let options = command.arguments[0] as? [String: Any] else {
            return nil
        }
        return options[key]
    }

    private func getBoolParameter(_ command: CDVInvokedUrlCommand, _ key: String) -> Bool? {
        return getParameter(command, key) as? Bool
    }
}

File: packages/cordova-plugin/android/OSFileTransferPlugin.kt
package com.outsystems.plugins.filetransfer

import io.ionic.libs.ionfiletransferlib.IONFLTRController
import io.ionic.libs.ionfiletransferlib.model.IONFLTRDownloadOptions
import io.ionic.libs.ionfiletransferlib.model.IONFLTRProgressStatus
import io.ionic.libs.ionfiletransferlib.model.IONFLTRTransferHttpOptions
import io.ionic.libs.ionfiletransferlib.model.IONFLTRTransferResult
import io.ionic.libs.ionfiletransferlib.model.IONFLTRUploadOptions
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.PluginResult
import org.json.JSONArray
import org.json.JSONObject

class OSFileTransferPlugin : CordovaPlugin() {
    companion object {
        private const val PROGRESS_UPDATE_INTERVAL = 100L // 100ms between progress updates
        private const val DEFAULT_TIMEOUT_MS = 60000 // 60 seconds default timeout
    }

    private lateinit var controller: IONFLTRController
    private val ioScope by lazy { CoroutineScope(Dispatchers.IO) }
    private val listeners: MutableList<CallbackContext> = mutableListOf()
    private var lastProgressUpdate = 0L

    override fun pluginInitialize() {
        super.pluginInitialize()
        controller = IONFLTRController(cordova.context)
    }

    override fun onDestroy() {
        super.onDestroy()
        ioScope.cancel()
    }

    override fun execute(
        action: String,
        args: JSONArray,
        callbackContext: CallbackContext
    ): Boolean {
        return when (action) {
            "downloadFile" -> {
                val options = parseDownloadOptions(args)
                if (options == null) {
                    callbackContext.sendError(OSFileTransferErrors.invalidParameters)
                } else {
                    downloadFile(options, callbackContext)
                }
                true
            }
            "uploadFile" -> {
                val options = parseUploadOptions(args)
                if (options == null) {
                    callbackContext.sendError(OSFileTransferErrors.invalidParameters)
                } else {
                    uploadFile(options, callbackContext)
                }
                true
            }
            "addListener" -> {
                listeners.add(callbackContext)
                true
            }
            "removeAllListeners" -> {
                callbackContext.sendPluginResult(PluginResult(PluginResult.Status.OK))
                listeners.forEach { 
                    // send an error to finish the plugin callback for "addListener"
                    it.sendPluginResult(PluginResult(PluginResult.Status.ERROR, "removeAllListeners was called"))
                }
                listeners.clear()
                true
            }
            else -> false
        }
    }

    private fun parseDownloadOptions(args: JSONArray): OSFileTransferDownloadOptions? {
        return try {
            val optionsObject = args.getJSONObject(0)
            val url = optionsObject.optString("url")
            val path = optionsObject.optString("path")
            
            if (url.isBlank() || path.isBlank()) {
                null
            }
            
            OSFileTransferDownloadOptions(
                url = url,
                path = path,
                progress = optionsObject.optBoolean("progress", false),
                httpOptions = parseHttpOptions(optionsObject, "GET")
            )
        } catch (e: Exception) {
            null
        }
    }

    private fun parseUploadOptions(args: JSONArray): OSFileTransferUploadOptions? {
        return try {
            val optionsObject = args.getJSONObject(0)
            val url = optionsObject.optString("url")
            val path = optionsObject.optString("path")
            
            if (url.isBlank() || path.isBlank()) {
                null
            }
            
            OSFileTransferUploadOptions(
                url = url,
                path = path,
                progress = optionsObject.optBoolean("progress", false),
                mimeType = optionsObject.optString("mimeType"),
                fileKey = optionsObject.optString("fileKey", "file"),
                chunkedMode = optionsObject.optBoolean("chunkedMode", false),
                httpOptions = parseHttpOptions(optionsObject, "POST")
            )
        } catch (e: Exception) {
            null
        }
    }

    private fun parseHttpOptions(options: JSONObject, defaultMethod: String): IONFLTRTransferHttpOptions {
        val headers = options.optJSONObject("headers") ?: JSONObject()
        val params = options.optJSONObject("params") ?: JSONObject()
        
        return IONFLTRTransferHttpOptions(
            method = options.optString("method", defaultMethod),
            headers = headers.toMap(),
            params = params.toParamsMap(),
            shouldEncodeUrlParams = options.optBoolean("shouldEncodeUrlParams", true),
            readTimeout = options.optInt("readTimeout", DEFAULT_TIMEOUT_MS),
            connectTimeout = options.optInt("connectTimeout", DEFAULT_TIMEOUT_MS),
            disableRedirects = options.optBoolean("disableRedirects", false)
        )
    }

    private fun JSONObject.toMap(): Map<String, String> {
        val map = mutableMapOf<String, String>()
        keys().asSequence().forEach { key -> 
            map[key] = optString(key, "")
        }
        return map
    }

    private fun JSONObject.toParamsMap(): Map<String, Array<String>> {
        val map = mutableMapOf<String, Array<String>>()
        this.keys().asSequence().forEach { key ->
            when (val value = this.opt(key)) {
                is String -> map[key] = arrayOf(value)
                is JSONArray -> {
                    val values = (0 until value.length())
                        .mapNotNull { value.optString(it, null) }
                        .toTypedArray()
                    if (values.isNotEmpty()) {
                        map[key] = values
                    }
                }
            }
        }
        return map
    }

    private fun downloadFile(options: OSFileTransferDownloadOptions, callbackContext: CallbackContext) {
        val downloadOptions = IONFLTRDownloadOptions(
            url = options.url,
            filePath = options.path,
            httpOptions = options.httpOptions
        )

        controller.downloadFile(downloadOptions)
            .onEach { result ->
                when (result) {
                    is IONFLTRTransferResult.Ongoing -> {
                        if (options.progress) {
                            notifyProgress("download", options.url, result.status)
                        }
                    }
                    is IONFLTRTransferResult.Complete -> {
                        // Send a final progress update with 100% completion
                        if (options.progress) {
                            val contentLength = result.data.totalBytes
                            val finalStatus = IONFLTRProgressStatus(
                                bytes = contentLength,
                                contentLength = contentLength,
                                lengthComputable = true
                            )
                            notifyProgress("download", options.url, finalStatus, forceUpdate = true)
                        }
                        
                        val response = JSONObject().apply {
                            put("path", options.path)
                        }
                        callbackContext.success(response)
                    }
                }
            }
            .catch { error ->
                val errorInfo = error.toOSFileTransferError().copy(
                    source = options.url,
                    target = options.path
                )
                callbackContext.sendError(errorInfo)
            }
            .launchIn(ioScope)
    }

    private fun uploadFile(options: OSFileTransferUploadOptions, callbackContext: CallbackContext) {
        val uploadOptions = IONFLTRUploadOptions(
            url = options.url,
            filePath = options.path,
            chunkedMode = options.chunkedMode,
            mimeType = options.mimeType,
            fileKey = options.fileKey,
            httpOptions = options.httpOptions
        )

        controller.uploadFile(uploadOptions)
            .onEach { result ->
                when (result) {
                    is IONFLTRTransferResult.Ongoing -> {
                        if (options.progress) {
                            notifyProgress("upload", options.url, result.status)
                        }
                    }
                    is IONFLTRTransferResult.Complete -> {
                        // Send a final progress update with 100% completion
                        if (options.progress) {
                            val contentLength = result.data.totalBytes
                            val finalStatus = IONFLTRProgressStatus(
                                bytes = contentLength,
                                contentLength = contentLength,
                                lengthComputable = true
                            )
                            notifyProgress("upload", options.url, finalStatus, forceUpdate = true)
                        }
                        
                        val response = JSONObject().apply {
                            put("bytesSent", result.data.totalBytes)
                            put("responseCode", result.data.responseCode)
                            put("response", result.data.responseBody)
                            
                            val headersObj = JSONObject()
                            result.data.headers?.entries?.forEach { (key, values) ->
                                key?.let { headerKey ->
                                    headersObj.put(headerKey, values.firstOrNull() ?: "")
                                }
                            }
                            put("headers", headersObj)
                        }
                        callbackContext.success(response)
                    }
                }
            }
            .catch { error ->
                val errorInfo = error.toOSFileTransferError().copy(
                    source = options.path,
                    target = options.url
                )
                callbackContext.sendError(errorInfo)
            }
            .launchIn(ioScope)
    }

    /**
     * Notify progress to listeners
     * Throttled to every 100ms to avoid excessive callbacks
     * 
     * @param transferType The type of transfer ("download" or "upload")
     * @param url The URL of the file being transferred
     * @param status The status of the transfer containing bytes, contentLength, etc.
     * @param forceUpdate If true, sends the update regardless of throttling
     */
    private fun notifyProgress(transferType: String, url: String, status: IONFLTRProgressStatus, forceUpdate: Boolean = false) {
        val currentTime = System.currentTimeMillis()
        if (forceUpdate || currentTime - lastProgressUpdate >= PROGRESS_UPDATE_INTERVAL) {
            val progressData = JSONObject().apply {
                put("type", transferType)
                put("url", url)
                put("bytes", status.bytes)
                put("contentLength", status.contentLength)
                put("lengthComputable", status.lengthComputable)
            }
            
            listeners.forEach { listener ->
                if (!listener.isFinished) {
                    val pluginResult = PluginResult(PluginResult.Status.OK, progressData)
                    pluginResult.keepCallback = true
                    listener.sendPluginResult(pluginResult)
                }
            }
            
            lastProgressUpdate = currentTime
        }
    }

    /**
     * Extension function to return an unsuccessful plugin result
     * @param error error class representing the error to return, containing a code and message
     */
    private fun CallbackContext.sendError(error: OSFileTransferErrors.ErrorInfo) {
        val errorJson = JSONObject().apply {
            put("code", error.code)
            put("message", error.message)
            if (error.source != null) put("source", error.source)
            if (error.target != null) put("target", error.target)
            if (error.httpStatus != null) put("httpStatus", error.httpStatus)
            if (error.body != null) put("body", error.body)
            
            if (error.headers != null) {
                val headersObj = JSONObject()
                error.headers.entries.forEach { (key, values) ->
                    key?.let { headerKey ->
                        headersObj.put(headerKey, values.firstOrNull() ?: "")
                    }
                }
                put("headers", headersObj)
            }
            
            if (error.exception != null) put("exception", error.exception)
        }
        
        val pluginResult = PluginResult(PluginResult.Status.ERROR, errorJson)
        this.sendPluginResult(pluginResult)
    }
}